src/openloco/ui/WindowManager.cpp

```diff
@@ -541,36 +579,294 @@ namespace openloco::ui::WindowManager
+                    loop = true;
+                    break;
+                }
+            }
+        } while (loop);
+
+        return foundSpace(type, position, size, flags, events);
     }
 
-    // 0x004C9F5D
+    /**
+     * 0x004C9F5D
+     *
+     * @param type @<cl>
+     * @param origin @<eax>
+     * @param size @<ebx>
+     * @param flags @<ecx << 8>
+     * @param events @<edx>
+     * @return
+     */
     window* createWindow(
         WindowType type,
         gfx::point_t origin,
         gfx::ui_size_t size,
         uint32_t flags,
         window_event_list* events)
     {
-        registers regs;
-        regs.eax = (origin.y << 16) | (origin.x & 0xFFFF);
-        regs.ebx = (size.height << 16) | (size.width & 0xFFFF);
-        regs.ecx = (uint8_t)type | (flags << 8);
-        regs.edx = (int32_t)events;
-        call(0x004C9F5D, regs);
-        return (window*)regs.esi;
+        console::log("Creating Window (%d, 0x%08X)", type, (uintptr_t)events);
+
+        if (count() == 12)
+        {
+            for (ui::window* w = &_windows[0]; w != _windowsEnd; w++)
+            {
+                if ((w->flags & window_flags::stick_to_back) != 0)
+                    continue;
+
+                if ((w->flags & window_flags::stick_to_front) != 0)
+                    continue;
+
+                if ((w->flags & window_flags::no_auto_close) != 0)
+                    continue;
+
+                close(w);
+                return createWindow(type, origin, size, flags, events);
+            }
+        }
+
+        // Find right position to insert new window
+        size_t dstIndex = 0;
+        if ((flags & window_flags::stick_to_back) != 0)
+        {
+            for (size_t i = 0; i < count(); i++)
+            {
+                if ((_windows[i].flags & window_flags::stick_to_back) != 0)
+                {
+                    dstIndex = i;
+                }
+            }
+        }
+        else if ((flags & window_flags::stick_to_front) != 0)
+        {
+            dstIndex = count();
+        }
+        else
+        {
+            for (int i = (int)count(); i > 0; i--)
+            {
+                if ((_windows[i - 1].flags & window_flags::stick_to_front) == 0)
+                {
+                    dstIndex = i;
+                    break;
+                }
+            }
+        }
+
+        auto window = ui::window(origin.x, origin.y, size.width, size.height);
+        window.type = type;
+        window.flags = flags;
+        if ((flags & window_flags::flag_12) != 0)
+        {
+            window.flags |= window_flags::white_border_mask;
+            audio::play_sound(audio::sound_id::open_window, origin.x + size.width / 2);
+        }
+        else if (((flags & window_flags::stick_to_back) == 0) && ((flags & window_flags::stick_to_front) == 0) && ((flags & window_flags::flag_13) == 0))
+        {
+            // FIXME: Same as previous case, but need a nice way to make the if clear
```

Comment:
Perhaps you could introduce temporary boolean variables, e.g. `has_flag_12`, `does_stick_to_back` to simplify the if-expression.

Reply:
The booleans have been introduced, but the two cases have yet to be merged.